File: dailyQuestion/2020-06/06-25/golang/solution_hash.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

// 删除重复的节点
func removeDuplicateNodes(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}
	tmp := map[int]int{head.Val: 0}
	pos := head
	for pos.Next != nil {
		cur := pos.Next
		if _, ok := tmp[cur.Val]; ok {
			pos.Next = pos.Next.Next
		} else {
			tmp[cur.Val] = 0
			pos = pos.Next
		}
	}
	return head
}

func main() {
	head := &ListNode{Val: 1}
	head.Next = &ListNode{Val: 2}
	head.Next.Next = &ListNode{Val: 3}
	head.Next.Next.Next = &ListNode{Val: 3}
	head.Next.Next.Next.Next = &ListNode{Val: 2}
	head.Next.Next.Next.Next.Next = &ListNode{Val: 1}

	result := removeDuplicateNodes(head)
	fmt.Println(result)
}

File: dailyQuestion/2020/2020-06/06-25/python/solution_hash.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def removeDuplicateNodes(head: ListNode) -> ListNode:
    if not head :
        return None
    tmp = [head.val]
    pos = head
    while pos.next:
        cur = pos.next
        if cur.val not in tmp:
            tmp.append(cur.val)
            pos = pos.next
        else:
            pos.next = pos.next.next
    return head

if __name__ == "__main__" :
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(3)
    head.next.next.next.next = ListNode(2)
    head.next.next.next.next.next = ListNode(1)

    result = removeDuplicateNodes(head)

    print(result)

